wdl-lint/src/v1/matching_parameter_meta.rs

```diff
@@ -0,0 +1,180 @@
+const ID: &str = "MatchingParameterMeta";
+
+/// Creates a "missing param meta" diagnostic.
+fn missing_param_meta(parent: &TaskOrWorkflow, missing: &str, span: Span) -> Diagnostic {
+    let (context, parent) = match parent {
+        TaskOrWorkflow::Task(t) => ("task", t.name()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name()),
+    };
+
+    Diagnostic::warning(format!(
+        "{context} `{parent}` is missing a parameter metadata key for input `{missing}`",
+        parent = parent.as_str(),
+    ))
+    .with_rule(ID)
+    .with_label(
+        "this input does not have an entry in the parameter metadata section",
+        span,
+    )
+    .with_fix(format!(
+        "add a `{missing}` key to the `parameter_meta` section with a detailed description of the \
+         input.",
+    ))
+}
+
+/// Creates an "extra param meta" diagnostic.
+fn extra_param_meta(parent: &TaskOrWorkflow, extra: &str, span: Span) -> Diagnostic {
+    let (context, parent) = match parent {
+        TaskOrWorkflow::Task(t) => ("task", t.name()),
+        TaskOrWorkflow::Workflow(w) => ("workflow", w.name()),
+    };
+
+    Diagnostic::warning(format!(
+        "{context} `{parent}` has an extraneous parameter metadata key named `{extra}`",
+        parent = parent.as_str(),
+    ))
+    .with_rule(ID)
+    .with_label(
+        "this key does not correspond to any input declaration",
+        span,
+    )
+    .with_fix("remove the extraneous parameter metadata entry")
+}
+
+/// Detects missing or extraneous entries in a `parameter_meta` section.
+#[derive(Debug, Clone, Copy)]
+pub struct MatchingParameterMetaRule;
+
+impl Rule for MatchingParameterMetaRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that inputs have a matching entry in a `parameter_meta` section."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Each input parameter within a task or workflow should have an associated `parameter_meta` \
+         entry with a detailed description of the input. Non-input keys are not permitted within \
+         the `parameter_meta` block."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Completeness])
+    }
+
+    fn visitor(&self) -> Box<dyn Visitor<State = Diagnostics>> {
+        Box::new(MatchingParameterMetaVisitor)
+    }
+}
+
+/// Checks for both missing and extra items in a `parameter_meta` section.
+fn check_parameter_meta(
+    parent: TaskOrWorkflow,
+    inputs: Option<InputSection>,
+    param_meta: Option<ParameterMetadataSection>,
+    diagnostics: &mut Diagnostics,
+) {
+    let expected: HashMap<_, _> = inputs
+        .iter()
+        .flat_map(|i| {
+            i.declarations().map(|d| {
+                let name = d.name();
+                (name.as_str().to_string(), name.span())
+            })
+        })
+        .collect();
+
+    let actual: HashMap<_, _> = param_meta
+        .iter()
+        .flat_map(|m| {
```

Comment:
I _think_ using a `flat_map` here is a problem. Meta can be arbitrarily nested, but we only want to check the first level. An example:
```
task {
    input { String foo }
    parameter_meta {
        not_foo: {
            also_not_foo: {
                bar: {
                    foo: "I might be describing some different `foo`!"
                }
            }
        }
    }
}
```
I guess the above would throw 3 warnings for `not_foo`, `also_not_foo`, and `bar`? But it would permit `foo`. I think instead We might only want to warn on the first level, (so ignore `also_not_foo` and `bar` _and_ `foo`).

Also I could be reading the Rust wrong lol.

Reply:
The code is only checking the top-level items in the parameter metadata.

The first `iter()` call is on the `param_meta` option, which will either yield the section or nothing.

As a result, the `flat_map` is effectively called at-most once, passing it the section. What's being flattened is the section's items and we don't look into the item's values, so it's only the top names that are iterated over.

I can add a test case to ensure we don't accidentally check for the nested names, though.

Reply:
Great, no code change needed then. But another test case would be a good sanity check 👍